pkg/controller/util/add_finalizer.go

```diff
@@ -0,0 +1,35 @@
+package util
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// AddFinalizer will add a finalizer to the PushApplication CR so that
+// we can delete from UPS appropriately
+func AddFinalizer(client client.Client, reqLogger logr.Logger, o metav1.Object) error {
+	// This is based on the example code at:
+	// https://github.com/operator-framework/operator-sdk/blob/master/doc/user-guide.md#handle-cleanup-on-deletion
+	if len(o.GetFinalizers()) < 1 && o.GetDeletionTimestamp() == nil {
+		reqLogger.Info("Adding Finalizer to the PushApplication")
+		o.SetFinalizers([]string{"finalizer.push.aerogear.org"})
+
+		// Update CR
+		switch typedObject := o.(type) {
+		case runtime.Object:
```

Comment:
This entire type switch is just shenanigans because the `metav1.Object` and `runtime.Object` interfaces for some reason are not aligned nicely.

Reply:
a type assertion should be better here IMO: https://tour.golang.org/methods/15

Reply:
```
        runtimeObject, ok := o.(runtime.Object)
	if !ok{
		reqLogger.Info("Can't determine the type of thing to add finalizer to")
		return nil
	}
```

Reply:
or, the entire method:

```
// AddFinalizer will add a finalizer to the PushApplication CR so that
// we can delete from UPS appropriately
func AddFinalizer(client client.Client, reqLogger logr.Logger, o metav1.Object) error {
	// This is based on the example code at:
	// https://github.com/operator-framework/operator-sdk/blob/master/doc/user-guide.md#handle-cleanup-on-deletion

	if len(o.GetFinalizers()) > 0 || o.GetDeletionTimestamp() != nil {
		return nil
	}

	reqLogger.Info("Adding Finalizer to the PushApplication")
	o.SetFinalizers([]string{"finalizer.push.aerogear.org"})

	runtimeObject, ok := o.(runtime.Object)
	if !ok {
		reqLogger.Info("Can't determine the type of thing to add finalizer to")
		return nil
	}

	err := client.Update(context.TODO(), runtimeObject)
	if err != nil {
		reqLogger.Error(err, "Failed to update a CR with a finalizer")
		return err
	}

	return nil
}
```

Reply:
Ah yes, that's nicer, I'll update it now, thanks!